### PR TITLE
Sudoku generator issue 3756

### PR DESCRIPTION
**Pull request by Radcliffe**
First version of Sudoku graph generator.

### CODE REVIEW

**Comment by Radcliffe** [comment]
Please review @ericmjl.

File: networkx/generators/tests/test_sudoku_graph.py [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+
+import pytest
+import networkx as nx
+
+
+def test_sudoku_negative():
+    """Verify that an error is raised when generating a Sudoku graph of order -1."""
+    pytest.raises(nx.NetworkXError, nx.sudoku_graph, n=-1)
+
+
```
Comment (ericmjl) [inline networkx/generators/tests/test_sudoku_graph.py]: I think it’s possible to make the tests a bit more concise by parameterizing them using pytest.

Here’s a strawman skeleton:

```python
@pytest.parametrize(“n”, [0, 1, 2, 3])
def test_sudoku_generator(n):
    G = nx.sudoku_graph(n)
    assert not G.is_directed()
    assert not G.is_multigraph()

    assert G.number_of_nodes() == n**4
    # not thinking straight right now, but probably putting in formula for number of edges is better
```

Since we’re doing a formal review here, I would highly recommend using this parametrized testing paradigm, rather than a series of example-based tests, because it then opens up the door to using property-based tests later (if NetworkX ever adopts [Hypothesis] for testing). 

[Hypothesis]: http://hypothesis.readthedocs.io/en/latest

@dschult and @jarrodmillman, would you be willing to chime in here?

File: networkx/generators/sudoku_graph.py [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+    >>> G.number_of_edges()
+    810
+    >>> sorted(G.neighbors(42))
+    [6, 15, 24, 33, 34, 35, 36, 37, 38, 39, 40, 41, 43, 44, 51, 52, 53, 60, 69, 78]
+    >>> G = nx.sudoku_graph(2)
+    >>> G.number_of_nodes()
+    16
+    >>> G.number_of_edges()
+    56
+
```
Comment (ericmjl) [inline networkx/generators/sudoku_graph.py]: Out of curiosity, what do you think of adding an example using the `equitable_color` function?

https://networkx.github.io/documentation/stable/reference/algorithms/generated/networkx.algorithms.coloring.equitable_color.html#networkx.algorithms.coloring.equitable_color

File: networkx/generators/tests/test_sudoku_graph.py (hunk 1 above)
Comment (Radcliffe) [reply networkx/generators/tests/test_sudoku_graph.py]: Thanks! I have parameterized the tests.

File: networkx/generators/sudoku_graph.py (hunk 2 above)
Comment (Radcliffe) [reply networkx/generators/sudoku_graph.py]: The `equitable_color` function yields a coloring with at least 21 colors, since the Sudoku graph has degree 20. The other coloring algorithms require fewer colors, but more than 9 in general. I would like to propose the addition of an exact minimum vertex coloring function, but this is a separate issue.

**Comment by dschult** [comment]
Just a quick comment about why the codecov test is complaining. I haven't looked at the code yet, but am interesting to see this be merged. 

The codecov complaint is that the new code (the patch) contains testing for only 90.62% of the new lines. Clicking on the details gives a report page from which you can select the "Diff" tab and it shows each line of the file colored "green" if that line has been tested, "red" if not tested, and "yellow" for statements (like if statements) that control the flow differently and only one of the two outcomes occurred during the tests.

In this case, the only lines untested are the last 3 lines of the file starting with ```if __name__ == "__main__"```. Those aren't needed since we are testing using the test files with pytest. So it should be enough to delete those lines.

**Comment by ericmjl** [comment]
@dschult I’ve given my approval on the PR. It’s thoroughly done, looks like it can withstand the evolution of the nx API long-term, and is well-tested.

**Comment by dschult** [comment]
This look good!  Thanks!

I found a couple of things to change:

- You should enter the new function into the docs by adding it to ```doc/reference/generators.rst``` (add the module and the function)

- Your module is named the same as your function causing one to hide the other. Perhaps make the module name ```sudoku_graphs.py```.  This also allows future functions involving other variants to be added to yours.

- The docstring for one of your tests extends to column 100 on a single line. ```Generate Sudoku graphs of various sizes and verify that they have the expected properties.```
I suggest rewording it to be shorter. Perhaps something like:
```Verify that various size Sudoku Graphs have expected properties.```

**Comment by dschult** [comment]
Hmmm...  I think this got merged with some other branch in a way that github now shows the unrelated changes as part of this PR instead of part of the previous PRs. It's better not to merge a PR with a new version of master. If you need/want new functionality try some form of ```git rebase```.

I think it'll all be cleaned up.  I ran the following after cloning your repository
```
git ch master
git remote add upstream https://github.com/networkx/networkx.git
git pull upstream master
git ch sudoku-generator-issue-3756
git rebase -i master
```

Then deleted all the commits listed there and replace with:
```
pick f8ad1fc Add Sudoku graph generator
pick 8fd1f41 Remove doctest and create_using parameter
pick a41e847 fix typos
pick 7fc1162 rewrite sudoku tests as a class
pick 47dd76b improve code style
pick a52dd11 fix typo in docstring
pick 7f78463 fix typo in docstring
pick a031dea Add reference
pick 530423c Parameterize tests for sudoku_graph
pick ea0c538 remove __main__
pick 11e2eb8 Reverting change to sudoku_graph.py
pick 08c58df remove __main__ from sudoku_graph.py
pick 3057a5b Improve documentation of sudoku submodule
```

Checked it with ```git diff master```
All your changes should be there and the changes from others should not appear.
Then pushed it to your repository and the PR should update just fine.

Can you just check the "changes" for this PR now and make sure I got everything?

**Comment by Radcliffe** [comment]
My apologies. I have always struggled with Git.

**Comment by ericmjl** [comment]
@dschult I just checked, looks like the relevant parts of the PR are still present and complete. @Radcliffe can you confirm?

**Comment by dschult** [comment]
Yay!  And Thanks!!  @Radcliffe

**Comment by Radcliffe** [comment]
Yay! Thank you all for being patient with a new contributor.
